crates/sui-core/src/authority_aggregator.rs

```diff
@@ -1011,201 +1007,135 @@ where
-                            error,
-                            state.signatures.iter().map(|s| s.authority).collect(),
-                            state.good_stake,
-                        ));
-                    }
-                }
+    ) -> SuiResult<Option<ProcessTransactionResult>> {
+        let (data, sig) = signed_transaction.into_inner().into_data_and_sig();
+        match state.tx_signatures.insert(sig) {
+            InsertResult::NotEnoughVotes => Ok(None),
+            InsertResult::Failed { error } => Err(error),
+            InsertResult::QuorumReached(cert_sig) => {
+                let ct = CertifiedTransaction::new_from_data_and_sig(data, cert_sig);
+                Ok(Some(ProcessTransactionResult::Certified(
+                    ct.verify(&self.committee)?,
+                )))
             }
         }
-        None
     }
 
-    fn handle_response_with_certified_transaction(
+    fn handle_transaction_response_with_executed(
         &self,
         state: &mut ProcessTransactionState,
         certificate: VerifiedCertificate,
         signed_effects: VerifiedSignedTransactionEffects,
-        name: AuthorityName,
-        weight: StakeUnit,
-    ) -> Option<ProcessTransactionResult> {
-        let tx_digest = certificate.digest();
+    ) -> SuiResult<Option<ProcessTransactionResult>> {
         // If we get a certificate in the same epoch, then we use it.
         // A certificate in a past epoch does not guarantee finality
         // and validators may reject to process it.
         if certificate.epoch() == self.committee.epoch {
-            debug!(?tx_digest, name=?name.concise(), weight, "Received prev certificate from validator handle_transaction");
-            return Some(ProcessTransactionResult::Certified(certificate));
-        } else if signed_effects.epoch() == self.committee.epoch {
+            Ok(Some(ProcessTransactionResult::Certified(certificate)))
+        } else {
             // If we get 2f+1 effects, it's an proof that the transaction
-            // has already been finalized in a different epoch. Regardless
-            // of the cert's epoch, we can accept it.
-            // This is safe when the signed-effects's epoch is equal to
-            // the local epoch because validators re-sign effects that are
-            // committed in past epochs. However it's not safe when the
-            // signed effects comes from the future because the stake
-            // distribution may have changed.
-            // Theoretically, the signed effects could be in a previous
-            // epoch from a stale validator, but in `effects_map` we try to
-            // form a CertifiedTransactionEffects which requires all sigs
-            // in the same epoch, this is not necessary but not a big deal
-            // anyways.
+            // has already been finalized. This works because validators would re-sign effects for transactions
+            // that were finalized in previous epochs.
```

Comment:
with finalized transactions, validators shouldn't need to re-sign txes finalized in a previous epoch. Though we can handle it in a separate PR to keep this one small

Reply:
Unfortunately validators will have to do so, at least for a while. There are two reasons:
1. If the information that validators return is not signed, we have no way to know the result is indeed from the validator, and could be prone to man-in-the-middle attacks. So the validator needs to sign the effects regardless.
2. Even if we implement authenticated communication, a client (e.g. the fullnode or wallet) will have no way to verify the information without a certified effects, until we have light client, which we don't plan to have before mainnet.

So I think at least before mainnet, we will stick to having validators resign effects

Reply:
+1

we have no good plan for any other form of finality proof other than an effects cert, and in order to always provide an effects cert, we need to continue re-signing.

I maintain there is nothing remotely weird about a validator attesting that a transaction was finalized in a previous epoch. The only reason it seems weird is because we overload the envelope epoch.